### PR TITLE
Tid processing enhancements

### PR DESCRIPTION
**Pull request by SerhiiNahornyi**

### CODE REVIEW

File: src/main/java/org/prebid/server/auction/requestfactory/VideoRequestFactory.java [hunk 1]
```diff
@@ -290,7 +290,12 @@ private WithPodErrors<BidRequest> fillImplicitParameters(HttpRequestContext http
                                                              boolean debugEnabled) {
 
         final BidRequest bidRequest = bidRequestToErrors.getData();
-        final BidRequest updatedBidRequest = paramsResolver.resolve(bidRequest, httpRequest, ENDPOINT);
+        final BidRequest updatedBidRequest = paramsResolver.resolve(
+                bidRequest,
+                httpRequest,
+
```
Comment (And1sS) [inline src/main/java/org/prebid/server/auction/requestfactory/VideoRequestFactory.java:296]: Redundant emptyline.

File: src/main/java/org/prebid/server/auction/requestfactory/Ortb2ImplicitParametersResolver.java [hunk 2]
```diff
@@ -410,9 +421,16 @@ private String getDomainOrNull(String url) {
...
     }
 
-    private Source populateSource(Source source, ExtRequest extRequest) {
+    private Source populateSource(Source source,
+                                  ExtRequest extRequest,
+                                  String endpoint,
+                                  boolean hasStoredBidRequest) {
         final String tid = source != null ? source.getTid() : null;
-        final String populatedTid = populateSourceTid(tid);
+        final String populatedTid = populateTidValue(tid,
```
Comment (And1sS) [inline src/main/java/org/prebid/server/auction/requestfactory/Ortb2ImplicitParametersResolver.java:429]: please, place `tid,` on next line.

File: src/main/java/org/prebid/server/auction/requestfactory/AuctionRequestFactory.java [hunk 3]
```diff
@@ -171,9 +172,17 @@ private Future<BidRequest> updateAndValidateBidRequest(AuctionContext auctionCon
...
+                .map(auctionStoredResult -> AuctionStoredResult.of(
+                        auctionStoredResult.isHasStoredBidRequest(),
+                        ortbVersionConversionManager.convertToAuctionSupportedVersion(
+                                auctionStoredResult.getBidRequest())
+                ))
 
-                .map(resolvedBidRequest -> paramsResolver.resolve(resolvedBidRequest, httpRequest, ENDPOINT))
+                .map(auctionStoredResult -> paramsResolver.resolve(auctionStoredResult.getBidRequest(),
+                                httpRequest,
+
```
Comment (And1sS) [inline src/main/java/org/prebid/server/auction/requestfactory/AuctionRequestFactory.java:183]: Redundant emptyline.

**Review by And1sS — CHANGES_REQUESTED** [review]
First round.

File: src/main/java/org/prebid/server/auction/StoredRequestProcessor.java [hunk 4]
```diff
@@ -109,7 +110,8 @@ private Future<BidRequest> processAuctionStoredRequest(String accountId, BidRequ
 
         return storedRequestsToBidRequest(
                 storedDataFuture, bidRequest, bidRequestToStoredRequestId.get(bidRequest), impToStoredRequestId)
-                .map(this::generateBidRequestIdForApp);
+                .map(this::generateBidRequestIdForApp)
+                .map(resolvedRequest -> AuctionStoredResult.of(true, resolvedRequest));
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/auction/StoredRequestProcessor.java:114]: Issue says that we need to generate `tid`
- for `BidRequest` if `request.ext.prebid.storedrequest.id` exists
- for `imp` if `imp.ext.prebid.storedrequest.id` exists.

Implemented logic doesn't match requirements.

File: src/main/java/org/prebid/server/auction/requestfactory/Ortb2ImplicitParametersResolver.java [hunk 5]
```diff
@@ -427,14 +445,29 @@ private Source populateSource(Source source, ExtRequest extRequest) {
...
+        final String generatedId = tidGenerator.generateId();
 
-        return StringUtils.isNotEmpty(generatedId) ? generatedId : null;
+        return StringUtils.isBlank(tid)
+                || (generateBidRequestId
+                && (org.prebid.server.model.Endpoint.openrtb2_amp.value().equals(endpoint) || hasStoredBidRequest))
+                && !containsTidMacro
+                ? generatedId
+                : StringUtils.replaceIgnoreCase(tid, OVERRIDE_SOURCE_ID_TEMPLATE, generatedId);
     }
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/auction/requestfactory/Ortb2ImplicitParametersResolver.java:471]: I think that `Endpoint.openrtb2_amp.value().equals(endpoint)` is redundant. We can always set `hasStoredBidRequest` as `true` for `amp`.

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/auction/StoredRequestProcessor.java (hunk 4 above)
Comment (SerhiiNahornyi) [reply src/main/java/org/prebid/server/auction/StoredRequestProcessor.java:114]: https://github.com/prebid/prebid-server/issues/2381 was updated

**Review by And1sS — APPROVED** [review]

File: src/main/java/org/prebid/server/auction/requestfactory/Ortb2ImplicitParametersResolver.java [hunk 6]
```diff
@@ -455,8 +492,22 @@ private List<Imp> populateImps(BidRequest bidRequest, HttpRequestContext httpReq
...
+                            imp,
+                            secureFromRequest,
+                            globalBidderParams,
+                            generateBidRequestId,
+                            hasStoredBidRequest,
+                            !isUniqueIds ? String.valueOf(index + 1) : null,
+                            mapper,
+                            tidGenerator,
+                            jsonMerger);
+                })
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/auction/requestfactory/Ortb2ImplicitParametersResolver.java:510]: `imps.get` can be inlined:
```
                .mapToObj(index -> new ImpPopulationContext(
                        imps.get(index),
                        secureFromRequest,
                        globalBidderParams,
                        generateBidRequestId,
                        hasStoredBidRequest,
                        !isUniqueIds ? String.valueOf(index + 1) : null,
                        mapper,
                        tidGenerator,
                        jsonMerger))
```

File: src/main/java/org/prebid/server/auction/requestfactory/Ortb2ImplicitParametersResolver.java [hunk 7]
```diff
@@ -762,23 +837,52 @@ public Imp getPopulationResult() {
...
+                    ? StringUtils.isNotBlank(impIdOverride) ? impIdOverride : generateSixteenDigitRandomString()
+                    : null;
+        }
+
+        private static String generateSixteenDigitRandomString() {
+            long smallest = 1000_0000_0000_0000L;
+            long biggest = 9999_9999_9999_9999L;
+            long random = ThreadLocalRandom.current().nextLong(smallest, biggest + 1);
+
+            return String.valueOf(random);
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/auction/requestfactory/Ortb2ImplicitParametersResolver.java:883]: What about to inline all of this?
```
return String.valueOf(
                    ThreadLocalRandom.current().nextLong(
                            1000_0000_0000_0000L,
                            1_0000_0000_0000_0000L));
```

**Review by CTMBNara — APPROVED** [review]

**Pull request merged**